### PR TITLE
Basic command created to gain experience

### PR DESCRIPTION
**Pull request by david22swan**

### CODE REVIEW

File: internal/pkg/pct/news.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+// The structure of the information we wish to retrieve
+type Item struct {
+	Title string `xml:"title" json:"title"`
+	Link  string `xml:"link" json:"link"`
+}
+
+func News(Url string, Format string) error {
+	fmt.Printf("Retrieving HTML code of %s ...\n", Url)
+	// Retrieve the html of the page
+	resp, err := http.Get(Url) //nolint:gosec
```
Comment (da-ar) [inline internal/pkg/pct/news.go:29]:
```suggestion
	resp, err := http.Get(Url) // #nosec
```

Comment (david22swan) [reply internal/pkg/pct/news.go]: Works :) Thank you

File: cmd/news/news.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	// If flag is invalid
+	if OutputFormat != "table" && OutputFormat != "json" {
+		return fmt.Errorf("Invalid value for format flag has been given")
+	}
+
+	return nil
+}
+
+func execute(cmd *cobra.Command, _ []string) error {
+	_ = pct.News(Url, OutputFormat)
```
Comment (da-ar) [inline cmd/news/news.go:76]: do not hide the error here with underscore, this is part of `RunE` which can handle errors, allow these to bubble up

File: internal/pkg/pct/news.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+		return err
+	}
+
+	// Create a variable of the top structure type set out above
+	var items RSS
+
+	// Unmarshal the xml
+	err = xml.Unmarshal([]byte(html), &items)
+	if err != nil {
+		return fmt.Errorf("Error ith retrieved yaml")
```
Comment (da-ar) [inline internal/pkg/pct/news.go:47]: Don't create error text when you have an error from a library. Always return the library error at a minimum
```suggestion
		return err
```

File: internal/pkg/pct/news.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	if Format == "table" {
+		outputAsTable(items)
+	} else if Format == "json" {
+		_ = outputAsJson(items)
```
Comment (da-ar) [inline internal/pkg/pct/news.go:53]: again, return and handle your errors

File: internal/pkg/pct/news.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	columnFmt := color.New(color.FgYellow).SprintfFunc()
+
+	// Add the titles to the table
+	tbl := table.New("Title", "Link")
+	tbl.WithHeaderFormatter(headerFmt).WithFirstColumnFormatter(columnFmt)
+
+	// Add each unmarshalled item as a new row
+	for i := 0; i < len(data.Item); i++ {
+		tbl.AddRow(data.Item[i].Title, data.Item[i].Link)
+	}
```
Comment (da-ar) [inline internal/pkg/pct/news.go:72]: This can be simplified by using `range` https://gobyexample.com/range

File: internal/pkg/pct/news.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+func outputAsJson(data RSS) error {
+	jsonData, _ := json.Marshal(data)
+
+	var cleanedJsonData bytes.Buffer
+	err := json.Indent(&cleanedJsonData, jsonData, "", "    ")
+	if err != nil {
+		return fmt.Errorf("Error formating Json")
+	}
```
Comment (da-ar) [inline internal/pkg/pct/news.go:85]: same as above, ensure that you return the errors from libraries

File: internal/pkg/pct/news_test.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var err error
+			_ = tt.args
+			if tt.expected.errorMsg != "" && err != nil {
+				assert.Contains(t, err.Error(), tt.expected.errorMsg)
+			} else {
+				assert.NoError(t, err)
+			}
+		})
```
Comment (da-ar) [inline internal/pkg/pct/news_test.go:60]: This isn't testing anything.  At no point do you call `pct.News` or `pct.outputAsTable` or `pct.outputAsJson`.

**Review by da-ar — CHANGES_REQUESTED** [review]
Missing tests for pct package

File: cmd/news/news.go (hunk 2 above)
Comment (david22swan) [reply cmd/news/news.go:76]: error is now captured and returned

File: internal/pkg/pct/news.go (hunk 3 above)
Comment (david22swan) [reply internal/pkg/pct/news.go:47]: err is nor returned

File: internal/pkg/pct/news.go (hunk 4 above)
Comment (david22swan) [reply internal/pkg/pct/news.go:53]: error is now captured and returned

File: internal/pkg/pct/news.go (hunk 5 above)
Comment (david22swan) [reply internal/pkg/pct/news.go:72]: Updated to:
```
for i := range(data.Item) {
    tbl.AddRow(data.Item[i].Title, data.Item[i].Link)
}
```

File: internal/pkg/pct/news.go (hunk 6 above)
Comment (david22swan) [reply internal/pkg/pct/news.go:85]: Updated

File: internal/pkg/pct/news_test.go (hunk 7 above)
Comment (da-ar) [reply internal/pkg/pct/news_test.go:60]: I've given a bad steer here. only `pct.News` can be tested while the output methods should be tested by passing the correct args into `pct.News`

Comment (david22swan) [reply internal/pkg/pct/news_test.go]: Have added proper unit test's that call the main News function

File: internal/pkg/pct/news.go (hunk 5 above)
Comment (da-ar) [reply internal/pkg/pct/news.go]: Range can return the value as well as the index. e.g.

```
for index, value:= range data.Item {
   tbl.AddRow(value.Title, value.Link)
}
```

As you can see with range you can skip the selection of `data.Item[i]` as the value of it is returned by range. In the above example you do not need to declare index, as it's not used, however i've written it there to show what range returns.

**Comment by da-ar** [comment]
Good work David

**Pull request closed without merging**